posthog/clickhouse/cancel.py

```diff
@@ -2,13 +2,44 @@
 
 from posthog.api.services.query import logger
 from posthog.clickhouse.client import sync_execute
+from posthog.clickhouse.client.connection import default_client
 from posthog.settings import CLICKHOUSE_CLUSTER
 
 
 def cancel_query_on_cluster(team_id: int, client_query_id: str) -> None:
-    result = sync_execute(
-        f"KILL QUERY ON CLUSTER '{CLICKHOUSE_CLUSTER}' WHERE query_id LIKE %(client_query_id)s",
-        {"client_query_id": f"{team_id}_{client_query_id}%"},
-    )
+    initiator_host = None
+    failed_to_find_initiator_host = False
+    try:
+        result = sync_execute(
+            f"""
+            SELECT hostname()
+            FROM clusterAllReplicas(posthog, system.query_log)
+            WHERE query_id LIKE %(client_query_id)s AND (event_time > (now() - toIntervalDay(1)))
+            GROUP BY hostname()
+            HAVING (count() = 1)
+            SETTINGS max_execution_time = 5
+            """,
+            {"client_query_id": f"{team_id}_{client_query_id}%"},
+        )
+        initiator_host = result[0][0] if result else None
+    except Exception as e:
+        failed_to_find_initiator_host = True
+        logger.info("Failed to find initiator host for query %s: %s", client_query_id, e)
+
+    if not failed_to_find_initiator_host:
+        if initiator_host:
+            logger.debug("Found initiator host for query %s, cancelling query on host", initiator_host, client_query_id)
+            with default_client(host=initiator_host) as client:
```

Comment:
Is it fine to use the `default_client` here?

My only concern is that the sync_execute method does much more to the query (setting a specific id, adding comments, and some additional stuff), but it's based on a Workload, instead of a specific node.

Reply:
This will lead to a bit of discrepancy, all sync_execute queries have a proper log_comment set, without it we won't be able to understand who initiated a kill. You can get it through a self join query_log table, but it makes queries ugly. Can add a setting with log_comment here?

Reply:
Yeah totally agree. I'll try to replicate what is done in sync_execute or even create another method that is similar but accepts the host.

Reply:
I changed this. Let me know your opinion, but now sync_execute is allowed to be passed a client, so I initialize it before passing it to the method.

That way we get all the benefits from sync_execute.

Reply:
LGTM, it's nice change.